### PR TITLE
Iss378

### PR DESCRIPTION
**Pull request by bechenard**
Ready to merge

### CODE REVIEW

**Comment by JeremyMcCormick** [comment]
I don't think there should be a separate HcalStripHit class.  The existing HcalHit class should likely be modified with your changes.  It should work fine as long as the class version is incremented.  Was there some good reason not to do this?

**Comment by bechenard** [comment]
HI Jeremy,

Sorry for the delayed reply, I don't check that mailbox often. Since Nhan
wanted to keep the old scheme to cross-validate the results, we decided to
go with separate products and modules for the time being. The trick is that
it is annoying to recover HcalHits from HcalStripHits as we randomly sample
the background in the strips or full plane.

We will likely replace the HcalHit with the HcalStripHit in the future
(when Nhan is happy).

Is that ok for you?
Cheers,
Bertrand

**Comment by nhanvtran** [comment]
Confirming...

File: EventProc/src/HcalStripDigiProducer.cxx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+#include "DetDescr/HcalID.h"
+#include "Event/HcalStripHit.h"
+
+namespace ldmx {
+
+    const float HcalStripDigiProducer::FIRST_LAYER_ZPOS = 569.5;
+    const float HcalStripDigiProducer::LAYER_ZWIDTH = 60.;
+    const int HcalStripDigiProducer::NUM_HCAL_LAYERS = 15;
+    const float HcalStripDigiProducer::MEV_PER_MIP = 1.40;
+    const float HcalStripDigiProducer::PE_PER_MIP = 13.5 * 6. / 4.;
```
Comment (nhanvtran) [inline EventProc/src/HcalStripDigiProducer.cxx]: So looking back at old talks, this should be 13.5 not 13.5 * 1.5...can we fix this and also in HcalDigiProducer.cxx?

File: SimApplication/src/HcalSD.cxx [hunk 2]
```diff
@@ -43,23 +44,43 @@ namespace ldmx {
...
         int copyNum = aStep->GetPreStepPoint()->GetTouchableHandle()->GetHistory()->GetVolume(layerDepth_)->GetCopyNo();
-        int layer = copyNum % 1000;
         int section = copyNum / 1000;
+        int layer = copyNum % 1000;
+
+        //stripID: back Hcal, segmented along y direction for now every 10 cm -- alternate x-y in the future?
+        //         left/right side hcal: segmented along x direction every 10 cm
+        //         top/bottom side hcal: segmented along y direction every 10 cm
+        int stripID = -1;        
+        if (section==HcalSection::BACK) stripID = int( (localPosition.y()+scint->GetYHalfLength())/100.0);
```
Comment (nhanvtran) [inline SimApplication/src/HcalSD.cxx]: Isn't this 1cm strips, not 10cm strips?

**Comment by nhanvtran** [comment]
General comment: this will break the old layer-wise HCAL veto. 
The general scheme is sum over all simhits and aggregate them in the same detIDraw.  Then for a given detIDraw, make a hit.  Once we add the strips into the detIDraw, then one cannot do the same thing for layers.  

Should we update the layer-wise HCAL veto at the same time?

**Comment by awhitbeck** [comment]
@nhanvtran I think we should make the changes to the old digi producer in this PR.  I think the simplest fix is to change `detIDraw` to `layer` for the map keys.   But I think this just effects the digi producers, not the veto processor.

@bechenard, I have a few minor comments for the `HcalStripDigiProducer`, some also apply equally to the `HcalDigiProducer`, which I can take care of, if you want:

 - This is wrong and probably should be removed:
https://github.com/LDMXAnalysis/ldmx-sw/blob/iss378/EventProc/src/HcalStripDigiProducer.cxx#L19-L21

 - Change layer to strip in comment: 
https://github.com/LDMXAnalysis/ldmx-sw/blob/iss378/EventProc/src/HcalStripDigiProducer.cxx#L83

 - Change var name to something more vague, or change layer to strip:
https://github.com/LDMXAnalysis/ldmx-sw/blob/iss378/EventProc/src/HcalStripDigiProducer.cxx#L83

I am still running some test, so there might be more to come, but after a quick look, it seems that everything for `HCalStripDigiProducer` works as expected.

**Comment by nhanvtran** [comment]
@awhitbeck yes I agree, we should put the fix in this PR
@awhitbeck for your 3rd point, I think the link is wrong

**Comment by awhitbeck** [comment]
I ran a test by generating muons which were randomly distributed in the x-y plane of the hcal.  I am seeing half the number of hits I would expect with strip id=30.  A plot is included which shows the hit multiplicity for all cells in the back section. 

image

**Comment by bechenard** [comment]
Weird I'll look into it. What are the bounds you use?

**Comment by awhitbeck** [comment]
If I understand the code, its -1500,1500 mm in both x and y directions.

**Comment by bechenard** [comment]
Hi Nhan,

Now, this is actually the correct definition. The strip 0 goes from -1550
to -1450, and if I have a point at -1500, then
localPosition.y()+scint->GetYHalfLength() = 50,
so dividing by 100 is actually the correct thing to do (dividing by 10
gives 1 cm strips).

**Comment by bechenard** [comment]
Hi Andrew,

For some (weird) reason, this has been coded as -1550 -> 1550. This might
explain the deficit, but I'll run more checks.

File: EventProc/src/HcalDigiProducer.cxx [hunk 3]
```diff
@@ -12,108 +12,130 @@
...
+            hit->setYpos(hcalYpos[detIDraw]);
+            hit->setZpos(hcalZpos[detIDraw]);
             ihit++;
-        } // end loop over detIDs (layers)
-        // put it into the event
-        event.add("hcalDigis", hits_);
+        } 
+        
+        
+        event.add("HcalDigis", hits_);
```
Comment (nhanvtran) [inline EventProc/src/HcalDigiProducer.cxx]: bug here...this should be `hcalDigis`, not `HcalDigis`

**Comment by nhanvtran** [comment]
Besides the one bug above, it looks great.  I would add some example configuration for usres in
`Configuration/python/sim_recon_prod.py.tmpl`

The default values should be:
```
hcalDigis.parameters["meanNoise"] = 1.5
hcalDigis.parameters["mev_per_mip"] = 1.4
hcalDigis.parameters["pe_per_mip"] = 13.5
hcalDigis.parameters["doStrip"] = 0
```

**Comment by bechenard** [comment]
Hi Nhan,

Should doStrip=0 be the default? Maybe for now.
I've fixed the bug and default values, and added a script to show how to use it (Configuration/python/sample_recon_Hcal.py) for free!

**Comment by nhanvtran** [comment]
Sure, @bechenard thanks, everything looks good on my side.
@awhitbeck did you want to run another muon spray validation?  or should we merge this guy?
@omar-moreno @JeremyMcCormick is that ok?

**Comment by bechenard** [comment]
Let Andrew spray first😀andrew could you post your script to do so

Thanks
Bertrand

**Comment by awhitbeck** [comment]
Yep, I'll rerun the test and post a recipe.

**Comment by omar-moreno** [comment]
Go for it.

**Comment by awhitbeck** [comment]
Looks ready to go to me.

**Pull request merged**